alternative-2-subset2-generic.Rmd

```diff
@@ -13,25 +13,45 @@ knitr::opts_chunk$set(
 
 # R Iterator Ideas (`[[` + signal, optionally with `as.iterable()`)
 
-R doesn't currently have a mechanism for users to customize what happens
-when objects are passed to `for`. Below is a proposal for what such a
-mechanism might look like:
+This approach extends the behavior of `[[` to signal a custom condition
+when done:
 
--   User implements `[[`, and optionally `as.iterable()`.
+```{r}
+stop_out_of_bounds <- function(call = sys.call()) {
+  cnd <- structure(
+    list(message = "subscript out of bounds", call = call),
+    class = c("out_of_bounds", "error", "condition")
+  )
+  stop(cnd)
+}
+```
 
--   User is expected to call `stop(out_of_bounds_condition())` in `[[`
-    when the iterable is exhausted.
+Which we use from `[[`:
 
--   `for` repeatedly calls `x[[i]]` with incrementing `i` until the
-    condition is encountered.
+```{r, error = TRUE}
+# Quick and dirty implementation; real implementation would need to happen in C
```

Comment:
I think the presentation is a bit simpler if we assume that base `[[` just throws an error of the correct type.